File: server-py/services/settings_service.py
from __future__ import annotations

from typing import Any
from urllib.parse import urlparse

import repositories.settings_repository as repo
from services.encryption import encrypt, decrypt, mask_api_key
from schemas import SettingsInput, VisibleSettings


def _http_error(message: str, status: int) -> Exception:
    e = Exception(message)
    e.status_code = status
    return e


_DEFAULT_MODEL = "gpt-4o-mini"


def get() -> dict[str, Any]:
    """获取设置并脱敏 API 密钥后返回给前端"""
    raw = repo.get_all()
    api_key_plain = _decrypt_safe(raw.get("apiKey", ""))
    return VisibleSettings(
        api_url=raw.get("apiUrl", ""),
        api_key_masked=mask_api_key(api_key_plain),
        model_id=raw.get("modelId", ""),
        system_prompt=raw.get("systemPrompt", ""),
        thinking_mode=raw.get("thinkingMode", "false") == "true",
    ).model_dump_camel()


def get_ai_settings() -> dict[str, Any]:
    """返回含明文 API 密钥的设置，供 AI API 调用使用（不返回给前端）"""
    raw = repo.get_all()
    api_key = _decrypt_safe(raw.get("apiKey", ""))
    return {
        "apiUrl": raw.get("apiUrl", ""),
        "apiKey": api_key,
        "modelId": raw.get("modelId", _DEFAULT_MODEL),
        "systemPrompt": raw.get("systemPrompt", ""),
        "thinkingMode": raw.get("thinkingMode", "false") == "true",
    }


def save(input_data: SettingsInput) -> None:
    """保存设置：校验 URL → 加密 API 密钥 → 持久化"""
    if not input_data.api_url or not input_data.model_id:
        raise _http_error("apiUrl and modelId are required", 400)

    parsed = urlparse(input_data.api_url)
    if not parsed.scheme or not parsed.netloc:
        raise _http_error("apiUrl must be a valid URL", 400)

    if not input_data.api_key:
        raise _http_error("apiKey is required", 400)

    settings: dict[str, str] = {
        "apiUrl": input_data.api_url,
        "apiKey": encrypt(input_data.api_key),  # 加密后再存储，绝不保存明文
        "modelId": input_data.model_id,
        "systemPrompt": input_data.system_prompt or "",
        "thinkingMode": "true" if input_data.thinking_mode else "false",
    }
    repo.upsert_all(settings)


def _decrypt_safe(encrypted: str) -> str:
    """安全解密，解密失败时返回脱敏标记而非抛出异常"""
    if not encrypted:
        return ""
    try:
        return decrypt(encrypted)
    except Exception:
        return "****"

File: server/services/settingsService.ts
import * as settingsRepo from '../repositories/settingsRepository.js';
import * as endpointRepo from '../repositories/endpointRepository.js';
import { encrypt, decrypt, maskApiKey } from './encryption.js';
import { RawSettings, SettingsInput, AiSettings, VisibleSettings } from '../types.js';

// 获取脱敏后的设置（API Key 解密后重新掩码）
export function get(): VisibleSettings {
  const raw: RawSettings = settingsRepo.getAll();
  let apiKeyMasked = '';
  if (raw.apiKey) {
    try {
      apiKeyMasked = maskApiKey(decrypt(raw.apiKey));
    } catch {
      apiKeyMasked = '****';  // 解密失败（如密钥变更），显示掩码
    }
  }
  const activeEndpoint = endpointRepo.getActive();
  return {
    apiUrl: raw.apiUrl || '',
    apiKeyMasked,
    modelId: raw.modelId || '',
    systemPrompt: raw.systemPrompt || '',
    thinkingMode: raw.thinkingMode === 'true',
    memoryEnabled: raw.memoryEnabled === 'true',
    routingMode: raw.routingMode || 'auto',
    reactMaxIterations: parseInt(raw.reactMaxIterations || '5', 10),
    toolMaxRetries: parseInt(raw.toolMaxRetries || '5', 10),
    showReactSteps: raw.showReactSteps !== 'false',
    activeEndpointId: activeEndpoint?.id || null,
    activeEndpointName: activeEndpoint?.name || null,
  };
}

// 获取内部使用的 AI 设置（优先从激活端点读取，兜底旧 settings）
export function getAiSettings(): AiSettings {
  const activeEndpoint = endpointRepo.getActive();
  if (activeEndpoint) {
    let apiKey = '';
    if (activeEndpoint.apiKey) {
      try {
        apiKey = decrypt(activeEndpoint.apiKey);
      } catch {
        apiKey = '';
      }
    }
    const raw: RawSettings = settingsRepo.getAll();
    return {
      apiUrl: activeEndpoint.apiUrl,
      apiKey,
      modelId: activeEndpoint.modelId,
      apiType: activeEndpoint.apiType || 'openai-chat',
      systemPrompt: raw.systemPrompt || '',
      thinkingMode: raw.thinkingMode === 'true',
      memoryEnabled: raw.memoryEnabled === 'true',
      reactMaxIterations: parseInt(raw.reactMaxIterations || '5', 10),
      toolMaxRetries: parseInt(raw.toolMaxRetries || '5', 10),
      showReactSteps: raw.showReactSteps !== 'false',
    };
  }
  // 兜底：旧 settings 表
  const raw: RawSettings = settingsRepo.getAll();
  return {
    apiUrl: raw.apiUrl || '',
    apiKey: raw.apiKey ? decrypt(raw.apiKey) : '',
    modelId: raw.modelId || 'gpt-4o-mini',
    apiType: 'openai-chat',
    systemPrompt: raw.systemPrompt || '',
    thinkingMode: raw.thinkingMode === 'true',
    memoryEnabled: raw.memoryEnabled === 'true',
    reactMaxIterations: parseInt(raw.reactMaxIterations || '5', 10),
    toolMaxRetries: parseInt(raw.toolMaxRetries || '5', 10),
    showReactSteps: raw.showReactSteps !== 'false',
  };
}

// 保存设置：API Key 加密后写入，仅在有新 key 时更新
// 同时同步 apiUrl/apiKey/modelId 到激活端点（若存在）
export function save({ apiUrl, apiKey, modelId, systemPrompt, thinkingMode, memoryEnabled, routingMode, reactMaxIterations, toolMaxRetries, showReactSteps }: SettingsInput): void {
  const settings: Record<string, string> = {
    apiUrl,
    modelId,
    systemPrompt: systemPrompt || '',
    thinkingMode: thinkingMode ? 'true' : 'false',
    memoryEnabled: memoryEnabled ? 'true' : 'false',
    routingMode: routingMode || 'auto',
    reactMaxIterations: String(reactMaxIterations ?? 5),
    toolMaxRetries: String(toolMaxRetries ?? 5),
    showReactSteps: showReactSteps !== undefined ? String(showReactSteps) : 'true',
  };
  if (apiKey) {
    settings.apiKey = encrypt(apiKey);
  }
  settingsRepo.upsertAll(settings);

  // 同步到激活端点
  const activeEndpoint = endpointRepo.getActive();
  if (activeEndpoint) {
    const fields: Record<string, unknown> = {};
    if (apiUrl) fields.apiUrl = apiUrl;
    if (apiKey) fields.apiKey = encrypt(apiKey);
    if (modelId) fields.modelId = modelId;
    if (Object.keys(fields).length > 0) {
      endpointRepo.update(activeEndpoint.id, fields);
    }
  }
}
